manuals/fsapfsmount.1
.Dd June 26, 2026
.Dt FSAPFSMOUNT 1
.Os
.Sh NAME
.Nm fsapfsmount
.Nd mounts an Apple File System (APFS) container
.Sh SYNOPSIS
.Nm fsapfsmount
.Op Fl f Ar file_system_index
.Op Fl o Ar offset
.Op Fl p Ar password
.Op Fl r Ar recovery_password
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm fsapfsmount
is a utility to mount an Apple File System (APFS) container
.Pp
.Nm fsapfsmount
is part of the
.Nm libfsapfs
package.
.Nm libfsapfs
is a library to access the Apple File System (APFS) format
.Pp
.Ar source
an Apple File System (APFS) container
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl f Ar file_system_index
specify a specific file system or "all"
.It Fl h
shows this help
.It Fl o Ar offset
specify the container offset in bytes
.It Fl p Ar password
specify the password (or passphrase)
.It Fl r Ar recovery_password
specify the recovery password (or passphrase)
.It Fl v
verbose output to stderr, while fsapfsmount will remain running in the \
foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fsapfsmount -f 1 -p Password /dev/sda1
fsapfsmount 20181214
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr fsapfsinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfsapfs/issues
.Sh COPYRIGHT
Copyright (C) 2018-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* rename volume name to volume label ?

* handle transaction identifier mismatch

* fill attribute value_data_file_extents
* complete snapshot support
* extended attribute extent support
  * extent support, add to Python binding
  * move libfsapfs_internal_extended_attribute_get_data_stream to libfsapfs_attributes.[ch]
  * for extended attributes by name look up use native B-tree structures

* handle non-inline attribute symbolic link

* improve format support
** check for more compression methods
** Add resource fork support
** read zlib compressed resource fork data footer
** Check deflate compressed files edge case where first byte is 0x#f
** add support for solaris and bsdos device numbers

* add snapshot support
** add API function to retrieve creation time
** add API function to retrieve change time
** get volume from snapshot?

* B-tree
** Validate (expected) btree level to detect loops - added maximum recursion depth for now
* Validate checksum
** B-tree node
** Container key bag
** Volume key bag
** Reaper
** Object map
** Space manager
* Container superblock read file IO handle add block size support
* use container data block vector in read file IO handle
** Volume superblock
** Container map
** Container key bag?
* Handle file size of 0, mapped file size does not support 0 ?
* Container data handle add decryption support and use it to read container key bag?

library:
* check if data block vector allocated size matches size of file extents
* get_extent function, check inode flags and set extent is sparse flag

* change container read bootstap as advised in documentation
** read backup container superblocks

* API
** Add function to provide volume size
** Add function to provide volume date and time values

* optimization
** read inode and extended attributes at once
** cache uncompressed blocks

* clean up / restructure
** refactor libfsapfs_encryption_aes_key_unwrap to use encryption context?

Python bindings:
* add extents object?

Tools:
* fsapfsinfo tool
** add volume selection via GUID in path
** prompt user for password or recovery password when needed
** print extended attribute values
** handle \r in filename
** print hardlinks for -E option?
** "-E all" instead of using next_object_identifier walk leaf nodes

* fsapfsmount tool
** prompt user for password or recovery password when needed
** make symbolic links an absolute path which accounts for mount point
** add owner and group identifier support?
** add support for all volumes
** add volume selection support
** does fuse override file types such as socket?
** dokan add symbolic link support

Tests:
* add tests for container_key_bag
* add tests for volume_key_bag
* add tests for key_encrypted_key

20180907
* see `git log' for more recent change log
* Initial version based on libfsrefs and libfsntfs



fsapfstools/fsapfsmount.c
/*
 * Mounts an Apple File System (APFS) container.
 *
 * Copyright (C) 2018-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fsapfstools_getopt.h"
#include "fsapfstools_i18n.h"
#include "fsapfstools_libcerror.h"
#include "fsapfstools_libclocale.h"
#include "fsapfstools_libcnotify.h"
#include "fsapfstools_libfsapfs.h"
#include "fsapfstools_output.h"
#include "fsapfstools_signal.h"
#include "fsapfstools_unused.h"
#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"

mount_handle_t *fsapfsmount_mount_handle = NULL;
int fsapfsmount_abort                    = 0;

/* Signal handler for fsapfsmount
 */
void fsapfsmount_signal_handler(
      fsapfstools_signal_t signal FSAPFSTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fsapfsmount_signal_handler";

	FSAPFSTOOLS_UNREFERENCED_PARAMETER( signal )

	fsapfsmount_abort = 1;

	if( fsapfsmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     fsapfsmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fsapfsmount to mount an Apple File System (APFS) container.";

	fsapfstools_option_t options[ ] = {
		{ 'f', "file_system_index", "specify a specific file system or \"all\"" },
		{ 'h', NULL, "shows this help" },
		{ 'o', "offset", "specify the container offset in bytes" },
		{ 'p', "password", "specify the password (or passphrase)" },
		{ 'r', "recovery_password", "specify the recovery password (or passphrase)" },
		{ 'v', NULL, "verbose output to stderr, while fsapfsmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "container", "an Apple File System (APFS) container" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libfsapfs_error_t *error                     = NULL;
	system_character_t *option_file_system_index = NULL;
	system_character_t *option_offset            = NULL;
	system_character_t *option_password          = NULL;
	system_character_t *option_recovery_password = NULL;
	system_character_t *source                   = NULL;
	char *program                                = "fsapfsmount";
	system_integer_t option                      = 0;
	int number_of_options                        = (int) ( sizeof( options ) / sizeof( fsapfstools_option_t ) );
	int verbose                                  = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point              = NULL;
	int result                                   = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations fsapfsmount_fuse_operations;

	system_character_t *option_extended_options  = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                         = { program, NULL };
	struct fuse_args fsapfsmount_fuse_arguments  = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args fsapfsmount_fuse_arguments  = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *fsapfsmount_fuse_channel   = NULL;
#endif
	struct fuse *fsapfsmount_fuse_handle         = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS fsapfsmount_dokan_operations;
	DOKAN_OPTIONS fsapfsmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fsapfstools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fsapfstools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fsapfstools_output_version_fprint(
	 stdout,
	 program );

	if( fsapfstools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fsapfstools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fsapfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'f':
				option_file_system_index = optarg;

				break;

			case (system_integer_t) 'h':
				fsapfstools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'r':
				option_recovery_password = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fsapfstools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source container.\n" );

		fsapfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		fsapfstools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libfsapfs_notify_set_stream(
	 stderr,
	 NULL );
	libfsapfs_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &fsapfsmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_file_system_index != NULL )
	{
		if( mount_handle_set_file_system_index(
		     fsapfsmount_mount_handle,
		     option_file_system_index,
		     &error ) != 1 )
		{
			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );

			fprintf(
			 stderr,
			 "Unsupported file system index defaulting to: all.\n" );
		}
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     fsapfsmount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set container offset.\n" );

			goto on_error;
		}
	}
	if( option_password != NULL )
	{
		if( mount_handle_set_password(
		     fsapfsmount_mount_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
	if( option_recovery_password != NULL )
	{
		if( mount_handle_set_recovery_password(
		     fsapfsmount_mount_handle,
		     option_recovery_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set recovery password.\n" );

			goto on_error;
		}
	}
	if( mount_handle_open(
	     fsapfsmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source container\n" );

		goto on_error;
	}
	if( mount_handle_is_locked(
	     fsapfsmount_mount_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to unlock source container\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		fsapfsmount_fuse_arguments.argc = 0;
		fsapfsmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &fsapfsmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fsapfsmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fsapfsmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &fsapfsmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	fsapfsmount_fuse_operations.open       = &mount_fuse_open;
	fsapfsmount_fuse_operations.read       = &mount_fuse_read;
	fsapfsmount_fuse_operations.release    = &mount_fuse_release;
	fsapfsmount_fuse_operations.getxattr   = &mount_fuse_getxattr;
	fsapfsmount_fuse_operations.listxattr  = &mount_fuse_listxattr;
	fsapfsmount_fuse_operations.opendir    = &mount_fuse_opendir;
	fsapfsmount_fuse_operations.readdir    = &mount_fuse_readdir;
	fsapfsmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	fsapfsmount_fuse_operations.getattr    = &mount_fuse_getattr;
	fsapfsmount_fuse_operations.readlink   = &mount_fuse_readlink;
	fsapfsmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	fsapfsmount_fuse_handle = fuse_new(
	                           &fsapfsmount_fuse_arguments,
	                           &fsapfsmount_fuse_operations,
	                           sizeof( struct fuse_operations ),
	                           fsapfsmount_mount_handle );

	if( fsapfsmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          fsapfsmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	fsapfsmount_fuse_channel = fuse_mount(
	                            mount_point,
	                            &fsapfsmount_fuse_arguments );

	if( fsapfsmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	fsapfsmount_fuse_handle = fuse_new(
	                           fsapfsmount_fuse_channel,
	                           &fsapfsmount_fuse_arguments,
	                           &fsapfsmount_fuse_operations,
	                           sizeof( struct fuse_operations ),
	                           fsapfsmount_mount_handle );

	if( fsapfsmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          fsapfsmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_destroy(
	 fsapfsmount_fuse_handle );

	fuse_opt_free_args(
	 &fsapfsmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &fsapfsmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &fsapfsmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	fsapfsmount_dokan_options.Version    = DOKAN_VERSION;
	fsapfsmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	fsapfsmount_dokan_options.SingleThread = TRUE;
#else
	fsapfsmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		fsapfsmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		fsapfsmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	fsapfsmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	fsapfsmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	fsapfsmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	fsapfsmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	fsapfsmount_dokan_operations.CreateDirectory      = NULL;
	fsapfsmount_dokan_operations.Cleanup              = NULL;
	fsapfsmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fsapfsmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fsapfsmount_dokan_operations.WriteFile            = NULL;
	fsapfsmount_dokan_operations.FlushFileBuffers     = NULL;
	fsapfsmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fsapfsmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fsapfsmount_dokan_operations.FindFilesWithPattern = NULL;
	fsapfsmount_dokan_operations.SetFileAttributes    = NULL;
	fsapfsmount_dokan_operations.SetFileTime          = NULL;
	fsapfsmount_dokan_operations.DeleteFile           = NULL;
	fsapfsmount_dokan_operations.DeleteDirectory      = NULL;
	fsapfsmount_dokan_operations.MoveFile             = NULL;
	fsapfsmount_dokan_operations.SetEndOfFile         = NULL;
	fsapfsmount_dokan_operations.SetAllocationSize    = NULL;
	fsapfsmount_dokan_operations.LockFile             = NULL;
	fsapfsmount_dokan_operations.UnlockFile           = NULL;
	fsapfsmount_dokan_operations.GetFileSecurity      = NULL;
	fsapfsmount_dokan_operations.SetFileSecurity      = NULL;
	fsapfsmount_dokan_operations.GetDiskFreeSpace     = NULL;
	fsapfsmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fsapfsmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	fsapfsmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	fsapfsmount_dokan_operations.Cleanup              = NULL;
	fsapfsmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fsapfsmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fsapfsmount_dokan_operations.WriteFile            = NULL;
	fsapfsmount_dokan_operations.FlushFileBuffers     = NULL;
	fsapfsmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fsapfsmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fsapfsmount_dokan_operations.FindFilesWithPattern = NULL;
	fsapfsmount_dokan_operations.SetFileAttributes    = NULL;
	fsapfsmount_dokan_operations.SetFileTime          = NULL;
	fsapfsmount_dokan_operations.DeleteFile           = NULL;
	fsapfsmount_dokan_operations.DeleteDirectory      = NULL;
	fsapfsmount_dokan_operations.MoveFile             = NULL;
	fsapfsmount_dokan_operations.SetEndOfFile         = NULL;
	fsapfsmount_dokan_operations.SetAllocationSize    = NULL;
	fsapfsmount_dokan_operations.LockFile             = NULL;
	fsapfsmount_dokan_operations.UnlockFile           = NULL;
	fsapfsmount_dokan_operations.GetFileSecurity      = NULL;
	fsapfsmount_dokan_operations.SetFileSecurity      = NULL;
	fsapfsmount_dokan_operations.GetDiskFreeSpace     = NULL;
	fsapfsmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fsapfsmount_dokan_operations.Unmounted            = NULL;
	fsapfsmount_dokan_operations.FindStreams          = NULL;
	fsapfsmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &fsapfsmount_dokan_options,
	          &fsapfsmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Apple File System (APFS) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( fsapfsmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 fsapfsmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &fsapfsmount_fuse_arguments );
#endif
	if( fsapfsmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &fsapfsmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

